concent_api/core/views.py

```diff
@@ -1241,6 +1209,604 @@ def store_message_and_message_status(
+
+    provider  = Client.objects.get_or_create_full_clean(provider_public_key)
+    requestor = Client.objects.get_or_create_full_clean(requestor_public_key)
+
+    subtask = Subtask(
+        task_id         = task_id,
+        subtask_id      = subtask_id,
+        provider        = provider,
+        requestor       = requestor,
+        state           = state.name,
+        next_deadline   = datetime.datetime.fromtimestamp(next_deadline, timezone.utc),
+    )
+
+    add_messages_to_subtask(
+        subtask,
+        task_to_compute,
+        report_computed_task,
+        force_get_task_result,
+        ack_report_computed_task,
+        reject_report_computed_task,
+        subtask_results_accepted,
+        subtask_results_rejected,
+    )
+
+    subtask.full_clean()
+    subtask.save()
+    return subtask
+
+
+def update_file_status(
+    client_public_key: bytes,
+):
+    force_get_task_result_list = PendingResponse.objects.filter(
+        Q(subtask__requestor__public_key = b64encode(client_public_key)) | Q(subtask__provider__public_key = b64encode(client_public_key)),
+        subtask__state                   = Subtask.SubtaskState.FORCING_RESULT_TRANSFER.name,  # pylint: disable=no-member
+    )
+
+    for pending_reponse in force_get_task_result_list:
+        report_computed_task    = deserialize_message(pending_reponse.subtask.report_computed_task.data.tobytes())
+        file_transfer_token     = create_file_transfer_token(
+            report_computed_task,
+            client_public_key,
+        )
+        if get_file_status(file_transfer_token):
+            subtask         = pending_reponse.subtask
+            subtask.state   = Subtask.SubtaskState.RESULT_UPLOADED.name  # pylint: disable=no-member
+            subtask.full_clean()
+            subtask.save()
+            mark_message_as_delivered(pending_reponse)
+
+            store_message_inbox(
+                response_type       = PendingResponse.ResponseType.ForceGetTaskResultUpload,
+                client_public_key   = subtask.provider.public_key_bytes,
+                queue               = PendingResponse.PendingResponseType.Receive,
+                subtask             = subtask,
+            )
+            store_message_inbox(
+                response_type       = PendingResponse.ResponseType.ForceGetTaskResultUpload,
+                client_public_key   = subtask.requestor.public_key_bytes,
+                queue               = PendingResponse.PendingResponseType.Receive,
+                subtask             = subtask,
+            )
+
+
+def update_concent_states(
+    client_public_key: bytes,
+):
+    clients_subtask_list = Subtask.objects.filter(
+        Q(requestor__public_key = b64encode(client_public_key)) | Q(provider__public_key = b64encode(client_public_key)),
+        state__in               = Subtask.SUBTASK_ACTIVE_STATES,
+        next_deadline__lte      = timezone.now()
+    )
+
+    update_file_status(client_public_key)
```

Comment:
I think this should happen after `update_states_of_subtasks`, it might take longer and come with unpredictable results from cluster. If we change to use non-atomic transactions in the future then we can successfully update subtasks and fail on cluster later.

Reply:
It can not happen after update_states_of_subtask, becuase if Requestor ping receive to get `ForceGetTaskResultUpload` and file will be already correctly uploaded, Concent at first will change state of this subtask to 'FAILED' and then return `ForceGetTaskFailed` to Requestor instead of `ForceGetTaskResultUpload`

Reply:
So it should be before the query. `update_file_status()` does not even touch the results of this query so there's no reason not to do it before it.

Also, we really do need to have atomic `send/` and quick. We'll be eaten alive by race conditions here if we don't ensure that we never process the same subtask in multiple simultaneous requests.

Reply:
:+1: